Avoid deadlock in event start

xsyann opened the pull request:
When using streaming, the main event loop is blocked to await tokens, if we try to execute a function (`Message.send()` here) in the same event loop, it leads to a deadlock.

Adding a `new_loop` parameter to the `run_sync` function allows to create a new event loop in the current Thread and allows to send a Message during a stream.

ramnes reviewed (commented):
Good if that fixes an issue.

It looks like a very weird approach to me though, because:
1. It doesn't really make sense to run an event loop for just one function;
2. I don't understand, as a `run_sync()` user, how do I know if I should pass `new_loop=True` or `new_loop=False`? Am I supposed to try one, and if that doesn't work, try the other one? My point being that the event loop handling probably shouldn't be the responsibility of the API user.

xsyann commented:
> Good if that fixes an issue.
> 
> It looks like a very weird approach to me though, because:
> 
> 1. It doesn't really make sense to run an event loop for just one function;
> 2. I don't understand, as a `run_sync()` user, how do I know if I should pass `new_loop=True` or `new_loop=False`? Am I supposed to try one, and if that doesn't work, try the other one? My point being that the event loop handling probably shouldn't be the responsibility of the API user.

1. We don't really have the choice as we need to execute the fonction in the Thread (as the main event loop is blocked)
I agree that a better solution would have been to make the token generator async to avoid blocking the main event loop but it is not yet implemented in llama_index.

2. I changed it to a `main_thread` boolean and added some doc. As it has a default value, generally a user shouldn't have to deal with it, except in this particular use case.

willydouhard commented on src/chainlit/sync.py:
```diff
@@ -21,9 +21,15 @@
         return sync(co)
     else:
+        if not main_thread:
+            return asyncio.run(co)
```
why do we use asyncio.run here instead of sync()? Also the main_thread parameter naming might be confusing since we check `threading.current_thread() == threading.main_thread()` in the condition above

ramnes reviewed (approved):
I like `force_new_loop` better 👍
